### PR TITLE
Add Org and Workspace Run Tasks resources

### PR DESCRIPTION
**Pull request by glennsarti**
## Description

This PR adds the `data.tfe_workspace_run_task` and
`resource.tfe_workspace_run_task` resources to the provider. These
resources will be able to manage Workspace level Run Tasks.

This PR adds the `data.tfe_organization_run_task` and
`resource.tfe_organization_run_task` resources to the provider. These
resources will be able to manage Organizational level Run Tasks.

## Testing plan

N/A Included Acceptance Tests

## External links

- [API documentation](https://www.terraform.io/cloud-docs/api-docs/run-tasks)

## Output from acceptance tests

~~Acceptance TFC instance isn't setup for Run Tasks yet.  Manual tests against my local environment are detailed in the comments~~

It is is now!

### CODE REVIEW

**Comment by glennsarti** [comment]
Workspace Run Tasks

```
?       github.com/hashicorp/terraform-provider-tfe     [no test files]
=== RUN   TestAccTFEWorkspaceRunTaskDataSource_basic
--- PASS: TestAccTFEWorkspaceRunTaskDataSource_basic (28.94s)
PASS
ok      github.com/hashicorp/terraform-provider-tfe/tfe 29.535s
?       github.com/hashicorp/terraform-provider-tfe/version     [no test files]
```

```
?       github.com/hashicorp/terraform-provider-tfe     [no test files]
=== RUN   TestAccTFEWorkspaceRunTask_create
--- PASS: TestAccTFEWorkspaceRunTask_create (33.60s)
=== RUN   TestAccTFEWorkspaceRunTask_import
--- PASS: TestAccTFEWorkspaceRunTask_import (26.05s)
PASS
ok      github.com/hashicorp/terraform-provider-tfe/tfe 60.206s
?       github.com/hashicorp/terraform-provider-tfe/version     [no test files]
```

File: tfe/data_source_organization_run_task_test.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+	url          = "%s"
+	name         = "foobar-task-%d"
+	hmac_key     = "Password1"
+	enabled      = false
+}
+
+data "tfe_organization_run_task" "foobar" {
+	organization      = resource.tfe_organization.foobar.id
+	name              = "foobar-task-%d"
+	depends_on = [tfe_organization_run_task.foobar]
```
Comment (brandonc) [inline tfe/data_source_organization_run_task_test.go:57]: 👍

**Review by brandonc — COMMENTED** [review]
We'll need some docs created under `website/docs` for the new resources/data sources

File: tfe/data_source_organization_run_task_test.go (hunk 1 above)
Comment (glennsarti) [reply tfe/data_source_organization_run_task_test.go:57]: and some changelog love too.

**Comment by glennsarti** [comment]
Lint errors fixed...now for docs

File: tfe/resource_tfe_organization_run_task.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+	"log"
+	"strings"
+
+	tfe "github.com/hashicorp/go-tfe"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func resourceTFEOrganizationRunTask() *schema.Resource {
+	return &schema.Resource{
+
```
Comment (sebasslash) [inline tfe/resource_tfe_organization_run_task.go:14]: Whitespace

File: tfe/resource_tfe_workspace_run_task_test.go [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+		rs, ok := s.RootModule().Resources[n]
+		if !ok {
+			return fmt.Errorf("Not found: %s", n)
+		}
+
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("No instance ID is set")
+		}
+
+		if rs.Primary.Attributes["workspace_id"] == "" {
```
Comment (sebasslash) [inline tfe/resource_tfe_workspace_run_task_test.go:84]: Should we check for `task_id` as well?

File: tfe/resource_tfe_workspace_run_task_test.go [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+			return fmt.Errorf("No Workspace ID is set")
+		}
+
+		rt, err := tfeClient.WorkspaceRunTasks.Read(ctx, rs.Primary.Attributes["workspace_id"], rs.Primary.ID)
+		if err != nil {
+			return fmt.Errorf("error reading Workspce Run Task: %w", err)
```
Comment (sebasslash) [inline tfe/resource_tfe_workspace_run_task_test.go:90]:
```suggestion
			return fmt.Errorf("error reading Workspace Run Task: %w", err)
```
Minor spelling mistake

File: tfe/testing.go [hunk 5]
```diff
@@ -50,6 +61,12 @@ func skipIfEnterprise(t *testing.T) {
 	}
 }
 
+func skipUnlessRunTasksDefined(t *testing.T) {
+	if _, ok := os.LookupEnv(RunTasksURLEnvName); !ok {
```
Comment (sebasslash) [inline tfe/testing.go:65]: `LookupEnv` will return `ok == true` even if `RUN_TASKS_URL=""`, so we should ensure the string is not empty before we write off the url as defined.

File: tfe/run_task_helpers_test.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+		t.Run(name, func(t *testing.T) {
+			got, err := fetchWorkspaceRunTask(test.taskName, test.workspace, test.org, client)
+
+			if (err != nil) != test.err {
+				t.Fatalf("expected error is %t, got %v", test.err, err)
+			}
+
+			if test.expectExists {
+				if got == nil || got.RunTask.Name != test.taskName {
+					t.Fatalf("wrong result\ngot: %#v\nwant: %#v", got, nil)
```
Comment (sebasslash) [inline tfe/run_task_helpers_test.go:129]: Was not aware of the `%#` formatting directive which includes type information! Really cool & useful!

**Review by sebasslash — CHANGES_REQUESTED** [review]
Overall things are looking good. I'm going to need to give it a second pass and make sure I didn't overlook any of the fine print! 😆 Hopefully I have given you enough to chew on! 

One really important thing to mention is to include the documentation pages as well!

File: tfe/resource_tfe_organization_run_task.go [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+			"organization": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"url": {
+				Type:     schema.TypeString,
+				Required: true,
```
[the remark this answers is not in the document]
Comment (glennsarti) [reply tfe/resource_tfe_organization_run_task.go:37]: Unless there's a pre-defined url validator within go-tfe or terraform itself (which I couldn't find) I would prefer to not add that to this PR. Validating URLs is notoriously picky and it's value quickly diminishes as false failures occur.

File: tfe/resource_tfe_workspace_run_task_test.go (hunk 3 above)
Comment (glennsarti) [reply tfe/resource_tfe_workspace_run_task_test.go:84]: I don't think so. We're checking that the Workspace Run Task Id exists, not that the relationships internal to TFC exist.

File: tfe/resource_tfe_workspace_run_task_test.go (hunk 4 above)
Comment (glennsarti) [reply tfe/resource_tfe_workspace_run_task_test.go:90]: Fixed

File: tfe/testing.go (hunk 5 above)
Comment (glennsarti) [reply tfe/testing.go]: Done.

File: tfe/resource_tfe_organization_run_task.go (hunk 2 above)
Comment (glennsarti) [reply tfe/resource_tfe_organization_run_task.go]: Fixed

**Comment by glennsarti** [comment]
> One really important thing to mention is to include the documentation pages as well!

Yup I know...

File: tfe/resource_tfe_organization_run_task.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (sebasslash) [reply tfe/resource_tfe_organization_run_task.go:36]: What do you think about using the validator in the std lib package `net/url`? 
```go
_, err := url.ParseRequestURI("https://my-run-task-host.com/")
if err != nil {
  // not a valid URL
}
```

[the remark this answers is not in the document]
Comment (glennsarti) [reply tfe/resource_tfe_organization_run_task.go]: Went with the inbuilt validator in `github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation`

ValidateFunc: validation.IsURLWithHTTPorHTTPS,

[the remark this answers is not in the document]
Comment (glennsarti) [reply tfe/resource_tfe_organization_run_task.go]: Added a commit + tests

File: tfe/resource_tfe_organization_run_task.go [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+	d.Set("category", task.Category)
+	d.Set("enabled", task.Enabled)
+	// The HMAC Key is always empty from the API so all we can do is
+	// echo the request's key to the response
+	d.Set("hmac_key", tfe.String(d.Get("hmac_key").(string)))
+
+	return nil
+}
+
+func resourceTFEOrganizationRunTaskImporter(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
```
[the remark this answers is not in the document]
Comment (glennsarti) [reply tfe/resource_tfe_organization_run_task.go:165]: Don't suppose we could do that as a separate PR. This PR is already quite large.

File: tfe/resource_tfe_organization_run_task.go [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+	task, err := fetchOrganizationRunTask(s[1], s[0], tfeClient)
+	if err != nil {
+		return nil, err
+	}
+
+	d.Set("organization", task.Organization.Name)
+	d.SetId(task.ID)
+
+	return []*schema.ResourceData{d}, nil
+}
```
[the remark this answers is not in the document]
Comment (glennsarti) [reply tfe/resource_tfe_organization_run_task.go:185]: Don't suppose we could do that as a separate PR. This PR is already quite large.

**Comment by glennsarti** [comment]
@sebasslash All ready for re-review

**Comment by glennsarti** [comment]
CI is now running the Run Task acceptance tests

e.g.
```
PASS tfe.TestAccTFEOrganizationRunTask_validateSchemaAttributeUrl (1.33s)
PASS tfe.TestAccTFEOrganizationRunTask_create (7.47s)
PASS tfe.TestAccTFEOrganizationRunTask_import (6.11s
```

**Comment by glennsarti** [comment]
Tests are green 💚

File: .circleci/config.yml [hunk 10]
```diff
@@ -42,6 +42,10 @@ jobs:
 
       - run:
           name: Acceptance Tests
+          environment:
+            # Note that https://httpstat.us/200 is enough to test CRUD operations on Run Tasks
+            # themselves BUT any Runs that use the Run Task will timeout and error
+            RUN_TASKS_URL: "https://httpstat.us/200"
```
Comment (sebasslash) [inline .circleci/config.yml:48]: 🔥 So awesome. Thanks for this Glenn

File: website/docs/r/organization_run_task.html.markdown [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+layout: "tfe"
+page_title: "Terraform Enterprise: tfe_organization_run_task"
+sidebar_current: "docs-resource-tfe-organization-run-task"
+description: |-
+  Manages Run tasks.
+---
+
+# tfe_organization_run_task
+
+[Run tasks](https://www.terraform.io/cloud-docs/workspaces/settings/run-tasks) allow Terraform Cloud to interact with external systems at specific points in the Terraform Cloud run lifecycle. Run tasks are reusable configurations that you can attach to any workspace in an organization
```
Comment (sebasslash) [inline website/docs/r/organization_run_task.html.markdown:11]: Period here as well.

File: website/docs/r/organization_run_task.html.markdown [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+The tfe_organization_run_task resource creates, updates and destroys Run tasks.
```
Comment (sebasslash) [inline website/docs/r/organization_run_task.html.markdown:13]: I would rephrase this. Typically we give a little more context about the resource if necessary. In this case, we should make the logical distinction between what a Workspace and Organization run task are.

**Review by sebasslash — CHANGES_REQUESTED** [review]
We're in the last stretch of the race! Implementation and testing looks really solid; really neat to see the go-tfe mocks in use.  Some minor things below on the documentation! Once addressed, this PR has my approval.

File: tfe/resource_tfe_organization_run_task.go [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+			"organization": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"url": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.IsURLWithHTTPorHTTPS,
```
[the remark this answers is not in the document]
Comment (glennsarti) [reply tfe/resource_tfe_organization_run_task.go:38]: Yup, behind the scenes it's just url.parse.  But I'd rather use a SDK helper :-)

File: website/docs/r/organization_run_task.html.markdown (hunk 11 above)
Comment (glennsarti) [reply website/docs/r/organization_run_task.html.markdown]: Fixed and found this in all my updates.

File: website/docs/d/workspace_run_task.html.markdown [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+layout: "tfe"
+page_title: "Terraform Enterprise: tfe_workspace_task"
+sidebar_current: "docs-datasource-tfe-workspace-task"
+description: |-
+  Get information on a Workspace Run task.
+---
+
+# Data Source: tfe_workspace_task
+
+[Run tasks](https://www.terraform.io/cloud-docs/workspaces/settings/run-tasks) allow Terraform Cloud to interact with external systems at specific points in the Terraform Cloud run lifecycle. Run tasks are reusable configurations that you can attach to any workspace in an organization
```
[the remark this answers is not in the document]
Comment (glennsarti) [reply website/docs/d/workspace_run_task.html.markdown]: Fixed

File: website/docs/r/organization_run_task.html.markdown (hunk 12 above)
Comment (glennsarti) [reply website/docs/r/organization_run_task.html.markdown]: So rather than that, I've adding links to our documentation as it's always up to date and there's a lot of context there which is too big for the provider docs.

**Review by sebasslash — APPROVED** [review]
🚀 🎸 🔥

File: tfe/resource_tfe_organization_run_task.go [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+	}
+
+	// Update the config.
+	d.Set("name", task.Name)
+	d.Set("url", task.URL)
+	d.Set("category", task.Category)
+	d.Set("enabled", task.Enabled)
+	// The HMAC Key is always empty from the API so all we can do is
+	// echo the request's key to the response
+	d.Set("hmac_key", tfe.String(d.Get("hmac_key").(string)))
```
Comment (brandonc) [inline tfe/resource_tfe_organization_run_task.go:160]: null [can also be returned](https://www.terraform.io/cloud-docs/api-docs/run-tasks#sample-response)... the tests may be hiding this since the provider default is "" ?

**Review by brandonc — CHANGES_REQUESTED** [review]
Just one question about hmac-key. I'm running through some smoke tests now

**Review by brandonc — COMMENTED** [review]
Smoke tested ✅

File: tfe/resource_tfe_organization_run_task.go [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+			"category": {
+				Type:     schema.TypeString,
+				Default:  "task",
+				Optional: true,
+			},
+
+			"hmac_key": {
+				Type:      schema.TypeString,
+				Sensitive: true,
+				Default:   "",
```
[the remark this answers is not in the document]
Comment (glennsarti) [reply tfe/resource_tfe_organization_run_task.go:50]: I've done some digging and I don't think I can.  The TF provider SDK for the TypeString _always_ returns a string. And seems to consider a nil/null value as `""` e.g. a zero or nil value for a TypeString is `""`

https://github.com/hashicorp/terraform-plugin-sdk/blob/main/helper/schema/schema.go#L2351

So end result is, even if I set `hmac_key => nil` or use the default as `nil`, it always converts that an empty string and there's nothing I can do about it.

This isn't the fault of the API or go-tfe, but rather TF internals.

[the remark this answers is not in the document]
Comment (glennsarti) [reply tfe/resource_tfe_organization_run_task.go:50]: The question is then asked... if TypeString is not correct, what type should I be using?  the answer is, I don't think there is.

File: tfe/resource_tfe_organization_run_task.go (hunk 15 above)
Comment (glennsarti) [reply tfe/resource_tfe_organization_run_task.go:160]: This is the same conversation as above.  Resolving for now.

**Review by brandonc — APPROVED** [review]
@glennsarti I think I made a mistake with how rails is handling this because hmac_key appears to be nil unless I set it to an actual string. Thanks for entertaining my ideas, nonetheless!
